File: LeetCode/DataStructure_II/Day15/105_construct_binary_tree_from_preorder_and_inorder_traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def construct_tree(self, inorder, parent, preorder):
        if len(preorder) == 0:
            return
        
        if len(preorder) == 1:
            curr = preorder.pop(0)
            node = TreeNode(curr)    
            return node
        
        curr = preorder.pop(0)
        node = TreeNode(curr)    
        
        index = -1
        for i in range(len(inorder)):
            if inorder[i] == curr:
                index = i
                break
                
        parent = node
        left = inorder[:index]
        right = inorder[index+1:]
        parent.left = self.construct_tree(left, parent, preorder[:index])
        parent.right = self.construct_tree(right, parent, preorder[index:])
        
        return parent
            
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        return self.construct_tree(inorder, None, preorder)
    
        

File: LeetCode/DayChallenges/DayChallenge_2022_July/105_Construct_Binary_Tree_from_Preorder_and_Inorder_Traversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {number[]} preorder
 * @param {number[]} inorder
 * @return {TreeNode}
 */
 var buildTree = function(preorder, inorder) {
    if(preorder.length === 0) return null;
    const root = new TreeNode(preorder[0]);
    const inorderIndex = inorder.findIndex(item => item === root.val);
    const inorderLeft = inorder.slice(0, inorderIndex);
    const inorderRight = inorder.slice(inorderIndex+1);
    const preorderLeft = preorder.slice(1, 1+inorderLeft.length);
    const preorderRight = preorder.slice(inorderLeft.length+1);  
    root.left = buildTree(preorderLeft, inorderLeft);
    root.right = buildTree(preorderRight, inorderRight);
    return root;
};